raiden-ts/src/channels/epics.ts

```diff
@@ -669,29 +676,45 @@ export const channelOpenEpic = (
         onchainSigner,
       );
 
-      return defer(() =>
-        tokenNetworkContract.functions.openChannel(
-          address,
-          partner,
-          action.payload.settleTimeout ?? settleTimeout,
-        ),
-      ).pipe(
-        assertTx('openChannel', ErrorCodes.CNL_OPENCHANNEL_FAILED, { log }),
-        // also retry txFailErrors: if it's caused by partner having opened, takeUntil will see
-        retryTx(provider.pollingInterval, undefined, txNonceErrors.concat(txFailErrors), { log }),
-        // if channel gets opened while retrying (e.g. by partner), give up to avoid erroring
-        takeUntil(
-          action$.pipe(
-            filter(channelOpen.success.is),
-            filter(
-              (action_) =>
-                action_.meta.tokenNetwork === tokenNetwork && action_.meta.partner === partner,
+      let deposit$: Observable<channelDeposit.request> = EMPTY;
+      if (action.payload.deposit?.gt?.(0))
+        // if it didn't fail so far, emit a channelDeposit.request in parallel with waitOpen=true
+        // to send 'approve' tx meanwhile we open the channel
+        deposit$ = of(
+          channelDeposit.request(
+            { deposit: action.payload.deposit, subkey: action.payload.subkey, waitOpen: true },
+            action.meta,
+          ),
+        );
+
+      return concat$(
+        deposit$,
+        defer(() =>
+          tokenNetworkContract.functions.openChannel(
+            address,
+            partner,
+            action.payload.settleTimeout ?? settleTimeout,
+          ),
+        ).pipe(
+          assertTx('openChannel', ErrorCodes.CNL_OPENCHANNEL_FAILED, { log }),
+          // also retry txFailErrors: if it's caused by partner having opened, takeUntil will see
+          retryTx(provider.pollingInterval, undefined, txNonceErrors.concat(txFailErrors), {
+            log,
+          }),
+          // if channel gets opened while retrying (e.g. by partner), give up to avoid erroring
+          takeUntil(
+            action$.pipe(
+              filter(channelOpen.success.is),
+              filter(
+                (action_) =>
+                  action_.meta.tokenNetwork === tokenNetwork && action_.meta.partner === partner,
+              ),
             ),
           ),
+          // ignore success so it's picked by channelEventsEpic
```

Comment:
Why do we not simply defer the deposit request here? Why do the complicated waiting stuff? Is there an  advantage? It feels like tightening the two epics too much together.

Reply:
Sorry, I don't get what you mean: this is exactly the same code as before (for open), the only addition is `concat`ing a `channelDeposit.request` before actually doing the channel open, so the `approve` can be sent/mined in parallel

Reply:
But isn't the approve done by the `makeDeposit` function which is called **after** the channel was opened? :thinking:

Reply:
No! This was previously like that, but this fixes that by ensuring we only wait before `setTotalDeposit` (notice we pass the `channelId$`, which is `mergeMapTo` after `approve`), and that's precisely the optimization here

Reply:
Okay something I don't understand. We have [this line](https://github.com/raiden-network/light-client/blob/aa7edf8560de85101707899e67d38fcb9990c553/raiden-ts/src/channels/epics.ts#L867). The `channel$` observable will emit only when the channel has become available (assuming the `waitOpen` was set). Only then the pipe can trigger the `makeDeposit` call. Not correct?
I'll give you the approval since everything else is fine, but I'm only too stupid to get this details here.

Reply:
No, `publishReplay` (fortunately, and that's why it was chosen here) is able to create the "inner" observable as soon as it's subscribed, and merge it to the output, which means `makeDeposit$` will get called right away, but the `channelId$` will only get a value once something goes through the filters, and if it happens before the inner subscription or is an error, the implicit `ReplaySubject(1)` will ensure this is also _cached_ and forwarded to the output.